pkg/operation/care/garden_health.go

```diff
@@ -0,0 +1,189 @@
+)
+
+var (
+	requiredGardenRuntimeManagedResources = sets.New(
+		etcd.Druid,
+		kubestatemetrics.ManagedResourceName,
+		gardensystem.ManagedResourceName,
+	)
+
+	requiredVirtualGardenManagedResources = sets.New(
+		resourcemanager.ManagedResourceName,
+		gardeneraccess.ManagedResourceName,
+		kubecontrollermanager.ManagedResourceName,
+	)
+)
+
+// GardenHealth contains information needed to execute health checks for garden.
+type GardenHealth struct {
+	garden          *operatorv1alpha1.Garden
+	gardenNamespace string
+	runtimeClient   client.Client
+	clock           clock.Clock
+}
+
+// NewHealthForGarden creates a new Health instance with the given parameters.
+func NewHealthForGarden(garden *operatorv1alpha1.Garden, runtimeClient client.Client, clock clock.Clock, gardenNamespace string) *GardenHealth {
+	return &GardenHealth{
+		runtimeClient:   runtimeClient,
+		garden:          garden,
+		clock:           clock,
+		gardenNamespace: gardenNamespace,
+	}
+}
+
+// CheckGarden conducts the health checks on all the given conditions.
+func (h *GardenHealth) CheckGarden(
+	ctx context.Context,
+	conditions []gardencorev1beta1.Condition,
+	thresholdMappings map[gardencorev1beta1.ConditionType]time.Duration,
+) []gardencorev1beta1.Condition {
+
+	var (
+		systemComponentsCondition        gardencorev1beta1.Condition
+		virtualGardenComponentsCondition gardencorev1beta1.Condition
+	)
+	for _, cond := range conditions {
+		switch cond.Type {
+		case operatorv1alpha1.GardenSystemComponentsHealthy:
+			systemComponentsCondition = cond
+		case operatorv1alpha1.VirtualGardenComponentsHealthy:
+			virtualGardenComponentsCondition = cond
+		}
+	}
+
+	checker := NewHealthChecker(h.runtimeClient, h.clock, thresholdMappings, nil, nil, nil, nil, nil)
+	newSystemComponentsCondition, err := h.checkGardenSystemComponents(ctx, checker, systemComponentsCondition)
+	newVirtualGardenComponentsCondition, err2 := h.checkVirtualGardenComponents(ctx, checker, virtualGardenComponentsCondition)
+
+	return []gardencorev1beta1.Condition{
+		NewConditionOrError(h.clock, systemComponentsCondition, newSystemComponentsCondition, err),
+		NewConditionOrError(h.clock, virtualGardenComponentsCondition, newVirtualGardenComponentsCondition, err2),
+	}
+}
+
+func (h *GardenHealth) checkGardenSystemComponents(
+	ctx context.Context,
+	checker *HealthChecker,
+	condition gardencorev1beta1.Condition,
+) (
+	*gardencorev1beta1.Condition,
+	error,
+) {
+	managedResources := sets.List(requiredGardenRuntimeManagedResources)
+	managedResources = append(managedResources, istio.ManagedResourceNames(true, "virtual-garden-")...)
+
+	if features.DefaultFeatureGate.Enabled(features.HVPA) {
+		managedResources = append(managedResources, hvpa.ManagedResourceName)
+	}
+	if h.garden.Spec.RuntimeCluster.Settings != nil &&
+		h.garden.Spec.RuntimeCluster.Settings.VerticalPodAutoscaler != nil &&
+		pointer.BoolDeref(h.garden.Spec.RuntimeCluster.Settings.VerticalPodAutoscaler.Enabled, false) {
```

Comment:
Would it make sense to extract the condition into a helper function to make the `if` condition more readable, similar to the one above?